File: Day4/day4.R
library(tidyverse)

input <- readLines("Day4/input.txt")

# part1 ####
check_within <- function(x) {
  map_lgl(x, function(x) {
    pos <- x %>% str_split("[-,]", simplify = T) %>%
      as.integer()
    overlap <- all(between(pos[1:2], pos[3],pos[4])) | all(between(pos[3:4], pos[1],pos[2]))
    overlap
  })
}
check_within(input) %>% sum()

# part2 ####
check_within2 <- function(x) {
  map_lgl(x, function(x) {
    pos <- x %>% str_split("[-,]", simplify = T) %>%
      as.integer()
    overlap <- any(between(pos[1:2], pos[3],pos[4])) | any(between(pos[3:4], pos[1],pos[2]))
    overlap
  })
}
check_within2(input) %>% sum()

File: Day4/day4.py
with open('Day4/input.txt') as file:
    lines = file.read().strip()

lines = lines.split("\n")
    
# PART 1
import re

def between(test, min, max):
  res = [min <= t <= max for t in test]
  return res

def check_within(line):
  cvals = re.findall('\\d+',line)
  vals = [int(x) for x in cvals]
  overlap = all(between(vals[0:2], vals[2],vals[3])) | all(between(vals[2:], vals[0],vals[1]))
  return overlap

sum([*map(check_within, lines)])

# PART 2

def check_within2(line):
  cvals = re.findall('\\d+',line)
  vals = [int(x) for x in cvals]
  overlap = any(between(vals[0:2], vals[2],vals[3])) | any(between(vals[2:], vals[0],vals[1]))
  return overlap

sum([*map(check_within2, lines)])

